src/Surfnet/StepupMiddleware/ApiBundle/Configuration/Service/AllowedSecondFactorListService.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/**
+ * Copyright 2017 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupMiddleware\ApiBundle\Configuration\Service;
+
+use Surfnet\Stepup\Configuration\Value\AllowedSecondFactorList;
+use Surfnet\Stepup\Configuration\Value\AllowedSecondFactorMap;
+use Surfnet\StepupMiddleware\ApiBundle\Configuration\Entity\ConfiguredInstitution;
+use Surfnet\StepupMiddleware\ApiBundle\Configuration\Repository\AllowedSecondFactorRepository;
+use Surfnet\StepupMiddleware\ApiBundle\Configuration\Repository\ConfiguredInstitutionRepository;
+
+final class AllowedSecondFactorListService
+{
+    /**
+     * @var AllowedSecondFactorRepository
+     */
+    private $allowedSecondFactorRepository;
+    /**
+     * @var ConfiguredInstitutionRepository
+     */
+    private $configuredInstitutionRepository;
+
+    public function __construct(
+        AllowedSecondFactorRepository $allowedSecondFactoryRepository,
+        ConfiguredInstitutionRepository $configuredInstitutionRepository
+    ) {
+        $this->allowedSecondFactorRepository   = $allowedSecondFactoryRepository;
+        $this->configuredInstitutionRepository = $configuredInstitutionRepository;
+    }
+
+    /**
+     * @return AllowedSecondFactorMap
+     */
+    public function getAllowedSecondFactorMap()
+    {
+        $allInstitutions = array_map(function (ConfiguredInstitution $configuredInstitution) {
+            return $configuredInstitution->institution;
+        }, $this->configuredInstitutionRepository->findAll());
+
+        $allowedSecondFactorMap = AllowedSecondFactorMap::mappedTo($allInstitutions);
```

Comment:
If the map is used, it must be implemented differently. Here a map is created that isn't actually a map, it is a list. It doesn't map anything. Furthermore, we now map lazily: the mapped value remains empty for institutions that allow all SecondFactorTypes. The empty value is converted to a `::ofTypes([])` rather than using `::blank()` and that is only done on demand when reading, not upon creation.

If you want to use it, the factory method should be something like:
```
AllowedSecondFactorMap::from($this->allowedSecondFactorRepository->findAll());
```

Alternative map (including probably a bunch of typo's as I typed it out here :wink: ):

```
class AllowedSecondFactorMap
{
    private $mappedAllowedSecondFactors;

    /**
     * @param AllowedSecondFactor[] $allowedSecondFactors
     */
    public static function from($allowedSecondFactors)
    {
        $allowedSecondFactorMap = new self();
        foreach ($allowedSecondFactors as $allowedSecondFactor) {
            $allowedSecondFactorMap->add($allowedSecondFactor);
        }

        return $allowedSecondFactorMap;
    }

    public function getAllowedSecondFactorListFor(Institution $institution)
    {
        $institution = strtolower($allowedSecondFactor->institution->getInstution());
        if (!array_key_exists($institution, $this->mappedAllowedSecondFactors)) {
            return AllowedSecondFactorList::blank();
        }

        return AllowedSecondFactorList::ofTypes($this->mappedAllowedSecondFactors[$institution]);
    }

    private function __construct()
    {
    }

    private function add(AllowedSecondFactor $allowedSecondFactor)
    {
        $institution = strtolower($allowedSecondFactor->institution->getInstution());

        $this->mappedAllowedSecondFactors[$institution][] = $allowedSecondFactor->secondFactor;
    }
}
```

Simpler, always correct (!), and uses the methods of other VO's as intended (blank 😉 )

Thoughts?